### PR TITLE
VideoCapture test to ensure that cv::VideoCapture does not share output buffers

### PR DESCRIPTION
**Pull request by asmorkalov**
Related work: https://github.com/opencv/opencv/pull/18377

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/videoio/test/test_video_io.cpp [hunk 1]
```diff
@@ -616,4 +616,45 @@ static vector<Ext_Fourcc_API> generate_Ext_Fourcc_API_nocrash()
...
+    Mat hardCopies[10];
+    for(int i = 0; i < 10; i++)
+    {
+        ASSERT_NO_THROW(cap >> frames[i]);
+        EXPECT_FALSE(frames[i].empty());
+        hardCopies[i] = frames[i].clone();
+    }
+
+    for(int i = 0; i < 10; i+=2)
+        frames[i].setTo(cv::Scalar(255,255,255));
```
Comment (alalek) [inline modules/videoio/test/test_video_io.cpp:646]: It is a bad idea to require changing of such buffers.
They can be allocated as read-only (any write attempt will raise a failure).
For example, OpenCL read-only "map" from GPU decoder.

---

Suggestion:
- remove this block
- rename test

Comment (asmorkalov) [reply modules/videoio/test/test_video_io.cpp]: I renamed the test and removed Mat modification as you suggested, but it's not what OpenCV users expect. Documentation does not provide any cases with read-only Mat object and a lot of examples do camera Mat modification.

**Comment by asmorkalov** [comment]
@alalek It's ready.

File: modules/videoio/test/test_video_io.cpp [hunk 2]
```diff
@@ -616,4 +616,42 @@ static vector<Ext_Fourcc_API> generate_Ext_Fourcc_API_nocrash()
...
+    Mat frames[10];
+    Mat hardCopies[10];
+    for(int i = 0; i < 10; i++)
+    {
+        ASSERT_NO_THROW(cap >> frames[i]);
+        EXPECT_FALSE(frames[i].empty());
+        hardCopies[i] = frames[i].clone();
+    }
+
+    for(int i = 1; i < 10; i++)
```
Comment (alalek) [inline modules/videoio/test/test_video_io.cpp:645]: i = 0

File: modules/videoio/test/test_video_io.cpp [hunk 3]
```diff
@@ -616,4 +616,42 @@ static vector<Ext_Fourcc_API> generate_Ext_Fourcc_API_nocrash()
...
+    {
+        Mat diff;
+        absdiff(frames[i], hardCopies[i], diff);
+        auto outilers = countNonZero(diff.reshape(1));
+        EXPECT_EQ(outilers, 0);
```
Comment (alalek) [inline modules/videoio/test/test_video_io.cpp:650]: One line check:
```
EXPECT_EQ(0, cv::norm(frames[i], hardCopies[i], NORM_INF)) << i;
```

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**